omod/src/main/java/org/openmrs/module/adminui/page/controller/account/ChangePasswordPageController.java

```diff
@@ -59,49 +63,104 @@ public String post(@MethodParam("getChangePassword") @BindParams ChangePassword
+		String GP_PASSWORD_CUSTOM_REGEX = Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_PASSWORD_CUSTOM_REGEX);
+		boolean GP_PASSWORD_REQUIRES_DIGIT = Boolean.parseBoolean(Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_PASSWORD_REQUIRES_DIGIT));
+		boolean GP_PASSWORD_REQUIRES_NON_DIGIT = Boolean.parseBoolean(Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_PASSWORD_REQUIRES_NON_DIGIT));
+		boolean GP_PASSWORD_REQUIRES_UPPER_AND_LOWER_CASE = Boolean.parseBoolean(Context.getAdministrationService().getGlobalProperty(OpenmrsConstants.GP_PASSWORD_REQUIRES_UPPER_AND_LOWER_CASE));
+
 		if (StringUtils.isBlank(changePassword.getOldPassword())) {
 			errors.rejectValue("oldPassword", "emr.account.changePassword.oldPassword.required",
-			    new Object[] { messageSourceService.getMessage("emr.account.changePassword.oldPassword.required") }, null);
+			    new Object[] { messageSourceService.getMessage("emr.account.changePassword.oldPassword.required") }, messageSourceService.getMessage("emr.account.changePassword.oldPassword.required"));
 		}
+
 		if (StringUtils.isBlank(changePassword.getNewPassword()) || StringUtils.isBlank(changePassword.getConfirmPassword())) {
 			errors.rejectValue(
-			    "newPassword",
-			    "emr.account.changePassword.newAndConfirmPassword.required",
-			    new Object[] { messageSourceService.getMessage("emr.account.changePassword.newAndConfirmPassword.required") },
-			    null);
+					"newPassword",
+					"emr.account.changePassword.newAndConfirmPassword.required",
+					new Object[] { messageSourceService.getMessage("emr.account.changePassword.newAndConfirmPassword.required") },
+					null);
 		} else if (!changePassword.getNewPassword().equals(changePassword.getConfirmPassword())) {
 			errors.rejectValue("", "emr.account.changePassword.newAndConfirmPassword.DoesNotMatch",
-			    new Object[] { messageSourceService
-			            .getMessage("emr.account.changePassword.newAndConfirmPassword.DoesNotMatch") }, null);
+					new Object[] { messageSourceService
+							.getMessage("emr.account.changePassword.newAndConfirmPassword.DoesNotMatch") }, messageSourceService
+							.getMessage("emr.account.changePassword.newAndConfirmPassword.DoesNotMatch"));
+		}
+
+		if (GP_PASSWORD_CANNOT_MATCH_USERNAME_OR_SYSTEMID && Context.getAuthenticatedUser().getUsername().equals(changePassword.getNewPassword())){
+			errors.rejectValue("newPassword", "emr.account.changePassword.newPassword.cantMatchUsername",
+					new Object[] { messageSourceService.getMessage("emr.account.changePassword.newPassword.cantMatchUsername") },
+					messageSourceService.getMessage("emr.account.changePassword.newPassword.cantMatchUsername") );
+		}
+
+		if (StringUtils.isNotBlank(GP_PASSWORD_CUSTOM_REGEX) && !GP_PASSWORD_CUSTOM_REGEX.matches(changePassword.getNewPassword())){
```

Comment:
This is not correct, the regex should be the argument value you pass in to the String.matches() method.
I would assume that you do something like this:

new RegexEvaluator(regex).evaluate()